packages/rpc-methods/src/restricted/notify.ts

```diff
@@ -0,0 +1,124 @@
+import {
+  PermissionSpecificationBuilder,
+  PermissionType,
+  RestrictedMethodOptions,
+  ValidPermissionSpecification,
+} from '@metamask/snap-controllers';
+import { NonEmptyArray } from '@metamask/snap-controllers/src/utils';
+import { ethErrors } from 'eth-rpc-errors';
+import { isPlainObject } from '../utils';
+
+const methodName = 'snap_notify';
+
+// Move all the types to a shared place when implementing more notifications
+export enum NotificationType {
+  native = 'native',
+}
+
+export type NotificationArgs = {
+  /**
+   * Enum type to determine notification type. Currently only supports 'native'.
+   */
+  type: NotificationType;
+
+  /**
+   * A message to show on the notification.
+   */
+  message: string;
+};
+
+export type NotifyMethodHooks = {
+  /**
+   * @param snapId - The ID of the Snap that created the notification.
+   * @param args - The notification arguments.
+   */
+  showNotification: (
+    snapId: string,
+    args: NotificationArgs,
+  ) => Promise<boolean>;
+};
+
+type SpecificationBuilderOptions = {
+  allowedCaveats?: Readonly<NonEmptyArray<string>> | null;
+  methodHooks: NotifyMethodHooks;
+};
+
+type Specification = ValidPermissionSpecification<{
+  permissionType: PermissionType.RestrictedMethod;
+  targetKey: typeof methodName;
+  methodImplementation: ReturnType<typeof getImplementation>;
+  allowedCaveats: Readonly<NonEmptyArray<string>> | null;
+}>;
+
+const specificationBuilder: PermissionSpecificationBuilder<
+  PermissionType.RestrictedMethod,
+  SpecificationBuilderOptions,
+  Specification
+> = ({ allowedCaveats = null, methodHooks }: SpecificationBuilderOptions) => {
+  return {
+    permissionType: PermissionType.RestrictedMethod,
+    targetKey: methodName,
+    allowedCaveats,
+    methodImplementation: getImplementation(methodHooks),
+  };
+};
+
+export const notifyBuilder = Object.freeze({
+  targetKey: methodName,
+  specificationBuilder,
+  methodHooks: {
+    showNotification: true,
+  },
+} as const);
+
+function getImplementation({ showNotification }: NotifyMethodHooks) {
+  return async function implementation(
+    args: RestrictedMethodOptions<[NotificationArgs]>,
+  ): Promise<boolean> {
+    const {
+      params,
+      context: { origin },
+    } = args;
+
+    return await showNotification(origin, getValidatedParams(params));
```

Comment:
I think we're now blocked here on some upcoming changes to the RateLimitController, but I think this method should return `null` if successful or an error otherwise. That's our approximation of a `void` method in JSON-RPC, which I think this should be.

Reply:
I synced this with the latest RateLimitController implementation and threw an error if `isRateLimited = true`

Reply:
We can merge this whenever `RateLimitController` API is finished, since we'll need code in the extension to wire it together